rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,490 @@
+}
+
+// EncryptionKeySet contains the list of active and rotated WrappedKeys for a
+// given usage.
+message EncryptionKeySet {
+  // ActiveKeys is a list of active, wrapped X25519 private keys. There should
+  // be at most one wrapped key per auth server using the
+  // SessionRecordingConfigV2 resource unless keys are being rotated.
+  repeated WrappedKey ActiveKeys = 1 [
+    (gogoproto.jsontag) = "active_keys"
+  ];
+  // RotatedKeys is a list of wrapped private keys that have been rotated.
+  // These are kept to decrypt historical encrypted session recordings.
+  repeated WrappedKey RotatedKeys = 2 [
+    (gogoproto.jsontag) = "rotated_keys"
+  ];
+}
+// SessionRecordingConfigStatusV2 contains all of the current and rotated keys
+// used for encrypted session recording.
+message SessionRecordingConfigStatusV2 {
+  EncryptionKeySet KeySet = 1 [(gogoproto.jsontag) = "keyset"]
+}
+
+// SessionRecordingConfigSpecV2 is the actual data we care about
+// for SessionRecordingConfig.
+message SessionRecordingConfigSpecV2 {
+  // existing fields omitted
+
+  SessionRecordingEncryptionConfig Encryption = 3 [
+    (gogoproto.jsontag) = "encryption"],
+    (gogoproto.nullable) = true,
+  ]
+}
+
+// SessionRecordingEncryptionConfig configures if and how session recordings
+// should be encrypted.
+message SessionRecordingEncryptionConfig {
+  BoolValue Enabled = 1 [(gogoproto.jsontag) = "encrypted"]
+}
+
+// SessionRecordingConfigV2 contains session recording configuration.
+message SessionRecordingConfigV2 {
+  // existing fields omitted
+
+  // Status contains all of the current and rotated keys used for encrypted
+  // session recording
+  SessionRecordingConfigStatusV2 Status = 6 [
+    (gogoproto.jsontag) = "status",
+    (gogoproto.nullable) = true
+  ]
+}
+```
+
+```proto
+// api/proto/teleport/clusterconfig/v1/clusterconfig_service.proto
+
+// existing messages omitted
+
+message RotateSessionRecordingConfigRequest {}
+message RotateSessionRecordingConfigResponse {}
+
+// ClusterConfigService provides methods to manage cluster configuration resources.
+service ClusterConfigService {
+  // existing RPCs omitted
+
+  // RotateSessionRecordingConfigKeys rotates the keys associated with encrypting and
+  // decrypting session recordings.
+  rpc RotateSessionRecordingConfig(RotateSessionRecordingConfigRequest) returns (RotateSessionRecordingConfigResponse);
+}
+```
+
+```proto
+// api/proto/teleport/legacy/client/proto/authservice.proto
+
+// existing messages omitted
+
+// EncryptedSessionRecordingChunk is an individual chunk of an encrypted 
+// session recording .tar file.
+message EncryptedSessionRecordingChunk {
+  // SessionID the recording relates to.
+  string SessionID = 1;
+  // ChunkIndex is the ordered index applied to the chunk.
+  int64 ChunkIndex = 2;
+  // Chunk is the encrypted chunk of session recording data being uploaded.
+  bytes Chunk = 3;
+}
+
+message UploadEncryptedSessionRecordingResponse {}
+
+service AuthService {
```

Comment:
New RPCs should be in separate services unless there's a good reason to keep them in the giant `proto.AuthService`.